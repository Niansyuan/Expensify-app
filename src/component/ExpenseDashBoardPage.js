import React from 'react'

const ExpenseDashBoardPage=()=>(
    <div>
        <p>This is from dashBoard component</p>
    </div>
)

export default ExpenseDashBoardPage