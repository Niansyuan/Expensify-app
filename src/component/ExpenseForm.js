import "react-dates/initialize";
import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? props.expense.amount.toString() : '',
            createAt: props.expense ? moment(props.expense.createAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }

    handleDiscriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }
    handleNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }
    handleAmountChange = (e) => {
        const amount = e.target.value;
        if (amount.match(/^\d{1,}(\.\d{0,2})?$/)) { //regular expression,詳看notion筆記
            this.setState(() => ({ amount }))
        }
    }
    handleDateChange = (createAt) => {
        this.setState(() => ({ createAt }))
    }
    handleFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }
    handleSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please Provide Description and Amount' }))
        } else {
            this.setState(() => ({ error: '' }))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10),
                createAt: this.state.createAt.valueOf(),
                note: this.state.note
            })
        }
    }
    render() {
        return (
            <form
                className="form"
                onSubmit={this.handleSubmit}
            >
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input
                    className="text-input"
                    type="text"
                    placeholder="Description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.handleDiscriptionChange}
                />
                <input
                    className="text-input"
                    type="number"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.handleAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createAt}
                    onDateChange={this.handleDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.handleFocusChange}
                    numberOfMonths={1} //一次顯示幾個月
                    isOutsideRange={() => false}
                />
                <textarea
                    className="text-area"
                    placeholder="Add a note for your expense...(optional)"
                    value={this.state.note}
                    onChange={this.handleNoteChange}
                >
                </textarea>
                <div>
                    <button className="button">Save Expense</button>
                </div>
            </form>
        )
    }
}