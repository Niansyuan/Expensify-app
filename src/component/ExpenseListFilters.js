import React from "react";
import { connect } from "react-redux";
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from "../actions/filters.js";
import { DateRangePicker } from "react-dates";

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };

    handleDateChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    handleFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    }
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }
    onSortChange = (e) => {
        if (e.target.value === "date") {
            this.props.sortByDate()
        };
        if (e.target.value === "amount") {
            this.props.sortByAmount()
        }
    }
    render() {
        return (
            <div className="contain-container">
                <div className="filter">
                    <div className="filter__contain">
                        <input
                            className="text-input"
                            type="text"
                            value={this.props.filters.text}
                            onChange={this.onTextChange}
                            placeholder='Search expenses' />
                    </div>
                    <div className="filter__contain">
                        <select
                            className="select"
                            value={this.props.filters.sortBy}
                            onChange={this.onSortChange}>
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="filter__contain">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            startDateId={"startDate"}
                            endDate={this.props.filters.endDate}
                            endDateId={"endDate"}
                            onDatesChange={this.handleDateChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.handleFocusChange}
                            numberOfMonths={1} //一次顯示幾個月
                            isOutsideRange={() => false}
                            showClearDates={true}
                        />
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
};
const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate())
})
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);