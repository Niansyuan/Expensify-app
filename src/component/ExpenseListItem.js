import React from "react";
import { Link } from 'react-router-dom';
import moment from "moment";
import numeral from "numeral";

const ExpenseListItem = ({ id, description, amount, createAt }) => {
    return (
        <Link
            to={`/edit/${id}`}
            className="list-item"
        >
            <div>
                <h3 className="list-item__title">{description}</h3>
                <span className="list-item__subtitle">{moment(createAt).format('MMM Do, YYYY')}</span>
            </div>
            <h3 className="list-item__data"> NT{numeral(amount).format('$0,0.00')}</h3>
        </Link>
    )
}

export default ExpenseListItem;