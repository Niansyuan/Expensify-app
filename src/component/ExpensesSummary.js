import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpenseTotal from '../selectors/expense-total';
import numeral from 'numeral';
import { Link } from 'react-router-dom';


export const ExpensesSummary = ({ expenseCount, expenseTotal }) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const expenseTotalFormat = numeral(expenseTotal).format('$0,0.00');
    return (
        <div className="page-header">
            <div className="contain-container">
                <h1 className="page-header__title">viewing <span>{expenseCount}</span> {expenseWord} totalling <span>NT{expenseTotalFormat}</span></h1>
                <div className="page-header__action">
                    <Link
                        to="/create"
                        className="button"
                    >Add Expense</Link>
                </div>
            </div>
        </div>
    )
}

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibleExpenses.length,
        expenseTotal: selectExpenseTotal(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary)