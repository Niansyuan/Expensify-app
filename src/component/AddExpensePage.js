import React from 'react'
import ExpenseForm from './ExpenseForm'
import { connect } from 'react-redux'
import { startAddExpenses } from '../actions/expenses'

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpenses(expense);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="contain-container">
                        <h1 className="page-header__title">Add Expense</h1>
                    </div>
                </div>
                <div className="contain-container">
                    <ExpenseForm
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>

        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpenses: (expense) => dispatch(startAddExpenses(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);