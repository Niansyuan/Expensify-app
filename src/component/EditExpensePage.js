import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm'
import { startEditExpenses, startRemoveExpenses } from '../actions/expenses'

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startEditExpenses(this.props.expense.id, expense)
        this.props.history.push('/')
        console.log('update', expense)
    }
    onClick = () => {
        this.props.startRemoveExpenses({ id: this.props.expense.id })
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="contain-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="contain-container">
                    <ExpenseForm
                        expense={this.props.expense}
                        onSubmit={this.onSubmit}
                    />
                    <div>
                        <button
                            className="button--remove"
                            expense={this.props.expense}
                            onClick={this.onClick}
                        >Remove Expense</button>
                    </div>
                </div>
            </div>
        )

    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch) => ({
    startEditExpenses: (id, expense) => dispatch(startEditExpenses(id, expense)),
    startRemoveExpenses: (data) => dispatch(startRemoveExpenses(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);