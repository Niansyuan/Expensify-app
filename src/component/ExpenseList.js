import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpense from "../selectors/expenses";

export const ExpenseList = (props) => (
    <div className="contain-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
            {
                props.expenses.length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>No expenses</span>
                    </div>
                ) : (
                    props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense} />;
                    })
                )
            }
        </div>
    </div>
)

//HOC -------
const mapStateToProps = (state) => {
    return {
        expenses: selectExpense(state.expenses, state.filters)
    }
}
export default connect(mapStateToProps)(ExpenseList);
//HOC -------