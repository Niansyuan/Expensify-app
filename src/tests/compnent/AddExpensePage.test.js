import React from 'react';
import { shallow } from 'enzyme';
import { AddExpensePage } from '../../component/AddExpensePage';
import "../setupTests";
import expense from '../fixture/expenses'

let startAddExpenses, history, wrapper;

beforeEach(() => {
    startAddExpenses = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddExpensePage startAddExpenses={startAddExpenses} history={history} />)
})

test('should render AddExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot();
})

test('should handle submit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expense[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startAddExpenses).toHaveBeenLastCalledWith(expense[1]);
})