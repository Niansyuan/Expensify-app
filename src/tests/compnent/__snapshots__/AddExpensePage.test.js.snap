// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpensePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="contain-container"
    >
      <h1
        className="page-header__title"
      >
        Add Expense
      </h1>
    </div>
  </div>
  <div
    className="contain-container"
  >
    <ExpenseForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
